### PR TITLE
Milestone rank

### PR DESCRIPTION
**Pull request by g4s8**
#575 - give higher rank to jobs with earlier 'due-date' in milestone.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#821](https://github.com/zerocracy/farm/issues/821) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#821](https://github.com/zerocracy/farm/issues/821) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/741359609). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/java/com/zerocracy/pm/staff/ranks/RnkMilestone.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    public RnkMilestone(final Github github) {
+        this.ghb = github;
+        this.cache = new HashMap<>(1);
+    }
+
+    @Override
+    public int compare(final String left, final String right) {
+        try {
+            return Long.compare(
+                this.dueDate(right).getTime(),
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/pm/staff/ranks/RnkMilestone.java]: @g4s8 `Date` is comparable. Why not `this.dueDate(right).compareTo(this.dueDate(left))`?

File: src/main/java/com/zerocracy/pm/staff/ranks/RnkMilestone.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Read due date from milestone.
+     * @param job Job id
+     * @return Mileston due-date
+     * @throws IOException If fails
+     */
+    private Date dueDate(final String job) throws IOException {
+        final Date date;
+        if (this.cache.containsKey(job)) {
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/pm/staff/ranks/RnkMilestone.java]: @g4s8 we can avoid the if-else statement altogether and simplify this by using [Map.computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-)
```java
    this.cache.computeIfAbsent(job, job -> {
            final Issue.Smart issue = new Issue.Smart(
                new Job.Issue(this.ghb, job)
            );
            if (issue.hasMilestone()) {
                date = new Milestone.Smart(issue.milestone()).dueOn();
            } else {
                date = RnkMilestone.MAX_DATE;
            }
            return date;
        }
    )
```

File: src/test/java/com/zerocracy/pm/staff/ranks/RnkMilestoneTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    @Test
+    @Ignore
+    public void giveHigherRankToEarlierDate() throws Exception {
+        final MkGithub github = new MkGithub();
+        final Repo repo = github.repos().create(
+            new Repos.RepoCreate("repos", false)
+        );
+        final Milestone release = repo.milestones().create("Release");
+        final Milestone pre = repo.milestones().create("PreRelease");
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pm/staff/ranks/RnkMilestoneTest.java]: @g4s8 let's call these variables `later` and `earlier` to be a bit clearer.

**Comment by 0crat** [comment]
@carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) this job was assigned to you 15days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
@carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) this job was assigned to you 15days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by carlosmiranda** [comment]
@g4s8 ping

**Comment by g4s8** [comment]
@carlosmiranda this PR will cause Github API quota issue again https://github.com/zerocracy/farm/issues/813
so I'm waiting for this bug to resolve: https://github.com/jcabi/jcabi-github/issues/1369
Maybe I need to close this PR and resubmit after jcabi-github fix?

**Comment by carlosmiranda** [comment]
@g4s8 yes, I think that would be a good idea.

**Comment by carlosmiranda** [comment]
@yegor256 PR rejected; it's currently outdated due to https://github.com/jcabi/jcabi-github/issues/1369.

**Comment by carlosmiranda** [comment]
@g4s8 can we close this now?

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#821](https://github.com/zerocracy/farm/issues/821) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request closed without merging**